### PR TITLE
jormungandr: synthese connector

### PR DESCRIPTION
**Pull request by TeXitoi**
http://jira.canaltp.fr/browse/NAVP-535

### CODE REVIEW

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    def __repr__(self):
+        """
+         used as the cache key. we use the rt_system_id to share the cache between servers in production
+        """
+        return self.rt_system_id
+
+    @cache.memoize(app.config['CACHE_CONFIGURATION'].get('TIMEOUT_SYNTHESE', 30))
+    def _call_synthese(self, url):
+        """
+        http call to timeo
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: old comments

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+        The call is handled by a circuit breaker not to continue calling timeo if the service is dead.
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: old comment

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            return None
+
+        if r.status_code != 200:
+            # TODO better error handling, the response might be in 200 but in error
+            logging.getLogger(__name__).error('Synthese RT service unavailable, impossible to query : {}'
+                                              .format(r.url))
+            return None
+
+        # TODO decode
+        logging.getLogger(__name__).info("synthese response: {}".format(r.text))
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:114]: .debug() no ?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        stop_point = str(route_point.fetch_stop_id(self.rt_system_id))
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: can't this be None ? why do you need str ?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        stop = route_point.fetch_stop_id(self.rt_system_id)
+
+        if not stop:
+            # one a the id is missing, we'll not find any realtime
+            logging.getLogger(__name__).debug('missing realtime id for {obj}: stop code={s}'.
+                                              format(obj=route_point, s=stop))
+            return None
+
+        stop_id_url = "SERVICE=tdg&roid={}".format(stop)
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: minor: can't you merge the 2 url building lines ?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+from jormungandr import cache, app
+from datetime import datetime, time
+
+class SyntheseRoutePoint(object):
+
+    def __init__(self, rt_id=None, sp_id=None):
+        self.syn_route_id = rt_id
+        self.syn_stop_point_id = sp_id
+
+    def __key(self):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:50]: why two '_'?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    def __repr__(self):
+        return "SyntheseRoutePoint({})".format(self.__key())
+
+
+class Synthese(RealtimeProxy):
+    """
+    class managing calls to timeo external service providing real-time next passages
+    """
+
+    def __init__(self, id, service_url, timezone, timeout=10):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:68]: :sob: this class will be broken with my PR: https://github.com/CanalTP/navitia/pull/1506

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            return None
+
+        if r.status_code != 200:
+            # TODO better error handling, the response might be in 200 but in error
+            logging.getLogger(__name__).error('Synthese RT service unavailable, impossible to query : {}'
+                                              .format(r.url))
+            return None
+
+        logging.getLogger(__name__).info("synthese response: {}".format(r.text))
+        stop_point = str(route_point.fetch_stop_id(self.rt_system_id))
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: `stop_point_id`

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        route = str(route_point.fetch_route_id(self.rt_system_id))
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: `route_id`

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            # one a the id is missing, we'll not find any realtime
+            logging.getLogger(__name__).debug('missing realtime id for {obj}: stop code={s}'.
+                                              format(obj=route_point, s=stop))
+            return None
+
+        url = "{base_url}?SERVICE=tdg&roid={stop_id}".format(base_url=self.service_url, stop_id=stop)
+
+        return url
+
+    def __get_value(self, item, xpath, val):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: two `_`?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        exceptions :
+            ValueError: Unable to parse datetime, day is out of range for month (for example)
+        '''
+        dt = date_time_format(xml_journey.get('dateTime'))
+        utc_dt = self.timezone.normalize(self.timezone.localize(dt)).astimezone(pytz.utc)
+        passage = RealTimePassage(utc_dt)
+        passage.is_real_time = (xml_journey.get('realTime') == 'yes')
+        return passage
+
+    def __build(self, xml):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: minor: this method can be a free function :)

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+            raise
+        for xml_journey in root.findall('journey'):
+            yield xml_journey
+
+    def _get_synthese_passages(self, xml):
+        result = {}
+        for xml_journey in self.__build(xml):
+            route_point = SyntheseRoutePoint(xml_journey.get('routeId'), self.__get_value(xml_journey, 'stop', 'id'))
+            if route_point not in result:
+                result[route_point] = []
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:173]: you can use `result = defaultdict(list)`

File: source/jormungandr/jormungandr/realtime_schedule/tests/synthese_xml_reader_test.py [hunk 13]
```diff
@@ -93,34 +94,40 @@ def get_xml_parser():
...
-    result = builder.get_synthese_passages(get_xml_parser())
+    builder = Synthese("id_synthese", "http://fake.url/", "Europe/Paris")
+    result = builder._get_synthese_passages(get_xml_parser())
     route_point = SyntheseRoutePoint('2533412229452279', '3377704015495922')
 
     assert route_point in result
     assert len(result[route_point]) == 1
     assert result[route_point][0].is_real_time == True
-    assert result[route_point][0].datetime == date_time_format("2016-Mar-21 12:07:37")
+    print "{}, {}".format(result[route_point][0].datetime, make_dt("2016-Mar-21 12:07:37"))
```
Comment (xlqian) [inline source/jormungandr/jormungandr/realtime_schedule/tests/synthese_xml_reader_test.py]: is this necessary?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 6 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:50]: copy/paste from internet (stack overflow I think)

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 8 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: That's not the scope of this PR to refactor the model ;-)

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 9 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: Same, not the scope

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 12 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:173]: But we'll have to loop 2 times on __build, no?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 10 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py]: Don't know, @azime ?

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 7 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:68]: I hope I'll be merged before you! ;-)

File: source/jormungandr/jormungandr/realtime_schedule/synthese.py (hunk 12 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/synthese.py:173]: Later, I need to know if there is a value (I return `m.get(route_point)`)

**Comment by xlqian** [comment]
Ok for me

**Pull request merged**